File: MapStat/stat_Cov.pl
#!/bin/perl
if(@ARGV!=3){
	print "\n  Usage:  perl $0 <Depth file> <RefGenome size> <Sample_name>\n\n";
	exit 1;
}

my $prefix=$ARGV[2];
my $total_size=$ARGV[1];

open DP,"$ARGV[0]";
my @depth_freq;
my $cov_size=0;
my $sum_depth=0;
while(<DP>){
	chomp;
	my @tmp_line=split("\t",$_);
	$depth_freq[$tmp_line[2]]+=1;
	$sum_depth+=$tmp_line[2];
	if($tmp_line[2]>0){$cov_size+=1;}
}
close DP;

open COV,">$prefix.Coverage.stat";
open DIST,">$prefix.Depth.freq";
open SUMMARY,">$prefix.Depth.stat";
open SUMMARY2,">$prefix.Depth2.stat";
print COV "Sample\tTotalBases\tCovBases\tCovRatio(%)\tAve_Depth\n";
print COV "$prefix\t$total_size\t$cov_size\t".sprintf("%.2f",$cov_size*100.0/$total_size)."\t".sprintf("%.2f",$sum_depth/$cov_size)."\n";
print DIST "Depth\tFreq\tRatio(%)\tCumulation(%)\n";
my $cumul_size=0;
my ($depth10_cov,$depth20_cov,$depth30_cov,$depth50_cov)=(0,0,0,0);
my $dp_percent80="";
for(my $i=1;$i<@depth_freq;$i+=1){
	if($depth_freq[$i] eq ""){$depth_freq[$i]=0;}
	$cumul_size+=$depth_freq[$i];
	my $cumul_ratio=sprintf("%.3f",$cumul_size*100.0/$cov_size);
	print DIST $i."\t".$depth_freq[$i]."\t".sprintf("%.3f",$depth_freq[$i]*100.0/$cov_size)."\t".$cumul_ratio."\n";
	if($dp_percent80 eq "" && $cumul_ratio >= 20){$dp_percent80 = $i; }
	if($i==9){
		$depth10_cov= sprintf("%.2f",(($cov_size-$cumul_size)*100.0/$total_size));
	}
	if($i==19){
		$depth20_cov= sprintf("%.2f",(($cov_size-$cumul_size)*100.0/$total_size));
	}
	if($i==29){
		$depth30_cov= sprintf("%.2f",(($cov_size-$cumul_size)*100.0/$total_size));
	}
	if($i==49){
		$depth50_cov= sprintf("%.2f",(($cov_size-$cumul_size)*100.0/$total_size));
	}

}
print SUMMARY "Sample\tAve_Depth(X)\t>=1X\t>=10X\t>=20X\t>=30X\t>=50X\n";
print SUMMARY "$prefix\t".sprintf("%.2f",$sum_depth/$cov_size)."\t".sprintf("%.2f",$cov_size*100.0/$total_size)."%\t".$depth10_cov."%\t".$depth20_cov."%\t".$depth30_cov."%\t".$depth50_cov."%\n";
close COV;
close DIST;
close SUMMARY;

print SUMMARY2 "Sample\tAve_Depth(X)\tFold80\t>=1X\t>=10X\t>=20X\t>=30X\t>=50X\n";
print SUMMARY2 "$prefix\t".sprintf("%.2f",$sum_depth/$cov_size)."\t".sprintf("%.3f",($sum_depth/$cov_size)/$dp_percent80)."\t".sprintf("%.2f",$cov_size*100.0/$total_size)."%\t".$depth10_cov."%\t".$depth20_cov."%\t".$depth30_cov."%\t".$depth50_cov."%\n";
close SUMMARY2;

File: MapStat/stat_Cov.py
#! /work/Software/Anaconda3-2019.10/bin/python

import sys
from collections import Counter, defaultdict
from functools import partial
from multiprocessing import Pool

import numpy as np
import pysam

doc = f"""
  Version:      v0.0.2 (2021-05-08)
  Authors:      SingChen
  Description:  Statistic Genome Coverage for WGS and WES

  Usage:        mosdepth -t 4 -x <Sample_name> <bam>
                {__file__} <per-base.bed> <RefGenome size> <Sample_name>
                
  Example:      {__file__} MS20041155P1.per-base.bed.gz 3095693983 MS20041155P1.Genome
"""


def calc_contig(contig, per_base, index):
    print("+", contig)
    cell = defaultdict(int)
    bed = pysam.TabixFile(per_base, index=index, parser=pysam.asTuple())
    for record in bed.fetch(contig):
        contig, start, end, depth = str(record).strip().split("\t")
        length = int(end) - int(start)
        cell[depth] += length
    bed.close()
    print("-", contig)
    return cell


def get_contigs(per_base, index):
    bed = pysam.TabixFile(per_base, index=index)
    contigs = bed.contigs
    bed.close()
    return contigs


def get_coverage_stat(counter, genome_size, sample, x1, avg_depth):
    out = sample + ".Coverage.stat"
    header = ["Sample", "TotalBases", "CovBases", "CovRatio(%)", "Ave_Depth"]
    row = [sample, str(genome_size), str(x1), f"{x1/genome_size*100:.2f}", f"{avg_depth:.2f}"]
    with open(out, "w") as f:
        f.write("\t".join(header) + "\n")
        f.write("\t".join(row) + "\n")


def get_depth_freq(counter, genome_size, sample, x1):
    out = sample + ".Depth.freq"
    header = ["Depth", "Freq", "Ratio(%)", "Cumulation(%)"]
    cum = 0
    depth_q20 = 0
    with open(out, "w") as f:
        f.write("\t".join(header) + "\n")
        for depth, length in sorted(counter.items(), key=lambda t: int(t[0])):
            if depth != "0":
                freq = length / x1 * 100
                cum += freq
                if not depth_q20 and cum >= 20:
                    depth_q20 = depth
                row = [depth, str(length), f"{freq:.3f}", f"{cum:.3f}"]
                f.write("\t".join(row) + "\n")
    return depth_q20


def get_depth_stat(counter, genome_size, sample):
    out = sample + ".Depth.stat"
    header = ["Sample", "Ave_Depth(X)", ">=1X", ">=10X", ">=20X", ">=30X", ">=50X", ">=100X", ">=150X", ">=200X"]
    cov_length = 0
    x0 = x1 = x10 = x20 = x30 = x50 = x100 = x150 = x200 = 0
    data = 0
    for depth, length in counter.items():
        depth = int(depth)
        if depth >= 200:
            x200 += length
            x150 += length
            x100 += length
            x50 += length
            x30 += length
            x20 += length
            x10 += length
            x1 += length
        elif depth >= 150:
            x150 += length
            x100 += length
            x50 += length
            x30 += length
            x20 += length
            x10 += length
            x1 += length
        elif depth >= 100:
            x100 += length
            x50 += length
            x30 += length
            x20 += length
            x10 += length
            x1 += length
        elif depth >= 50:
            x50 += length
            x30 += length
            x20 += length
            x10 += length
            x1 += length
        elif depth >= 30:
            x30 += length
            x20 += length
            x10 += length
            x1 += length
        elif depth >= 20:
            x20 += length
            x10 += length
            x1 += length
        elif depth >= 10:
            x10 += length
            x1 += length
        elif depth >= 1:
            x1 += length
        else:
            x0 += length
        data += depth * length
    avg_depth = data / x1
    row = [
        sample,
        f"{avg_depth:.2f}",
        f"{x1/genome_size*100:.2f}%",
        f"{x10/genome_size*100:.2f}%",
        f"{x20/genome_size*100:.2f}%",
        f"{x30/genome_size*100:.2f}%",
        f"{x50/genome_size*100:.2f}%",
        f"{x100/genome_size*100:.2f}%",
        f"{x150/genome_size*100:.2f}%",
        f"{x200/genome_size*100:.2f}%",
    ]
    with open(out, "w") as f:
        f.write("\t".join(header) + "\n")
        f.write("\t".join(row) + "\n")
    return x1, avg_depth


def get_cv_cov20(per_base, index, avg_depth, sample):
    array = []
    avg_p20 = avg_depth * 0.2
    small = large = 0
    bed = pysam.TabixFile(per_base, index=index, parser=pysam.asTuple())
    for record in bed.fetch():
        contig, start, end, depth = str(record).strip().split("\t")
        length = int(end) - int(start)
        depth = int(depth)
        for i in range(length):
            array.append(depth)
        if depth >= avg_p20:
            large += length
        else:
            small += length
    bed.close()
    cv = np.std(array) / np.mean(array)
    q20 = large / (large + small)
    return cv, q20


def main(per_base, genome_size, sample):
    genome_size = int(genome_size)
    counter = Counter()
    index = per_base + ".csi"
    contigs = get_contigs(per_base, index)
    func = partial(calc_contig, per_base=per_base, index=index)
    with Pool(8) as p:
        cells = p.map(func, contigs)
    for cell in cells:
        counter.update(cell)
    x1, avg_depth = get_depth_stat(counter, genome_size, sample)
    get_coverage_stat(counter, genome_size, sample, x1, avg_depth)
    depth_q20 = get_depth_freq(counter, genome_size, sample, x1)
    fold80 = avg_depth / int(depth_q20)
    cv, q20 = get_cv_cov20(per_base, index, avg_depth, sample)
    stat = sample + ".Depth.stat"
    out = sample + ".Depth-fold80.stat"
    with open(stat) as fi, open(out, "w+") as fo:
        fo.write(next(fi).strip() + "\tFold80\tCV\t>=20%X\n")
        fo.write(next(fi).strip() + f"\t{fold80:.4f}\t{cv:.2f}\t{q20*100:.2f}%\n")


if __name__ == "__main__":
    if len(sys.argv) != 4:
        sys.exit(doc)
    else:
        main(*sys.argv[1:])
